### PR TITLE
Enhance bag part to honor permissions

### LINKED ISSUE #10986: Improve BagPart
### Is your feature request related to a problem? Please describe.
When creating multiple similar items into a BagPart, it would be nice to have a clone function. 

### Describe the solution you'd like
Add a button to clone an item in a bag by adding a button next to the delete button.

Also, when there is only one item type that can be added, instead of showing `Add Item`, show `Add [ItemDisplayText]. This just saved 1 extra click X the amount of items to add.

The view will look something like this

```
@using OrchardCore.Flows.Models;
@using OrchardCore.Flows.ViewModels;
@using OrchardCore.Mvc.Utilities;

@model BagPartEditViewModel

@inject IContentManager ContentManager
@inject OrchardCore.ContentManagement.Metadata.IContentDefinitionManager ContentDefinitionManager
@inject OrchardCore.ContentManagement.Display.IContentItemDisplayManager ContentItemDisplayManager

@{
    var widgetTemplatePlaceholderId = Html.Id("widgetTemplatePlaceholder");
    var htmlFieldPrefix = ViewContext.ViewData.TemplateInfo.HtmlFieldPrefix;
    var parentContentType = Model.BagPart.ContentItem.ContentType;
    string partName = ((dynamic)Model).Metadata.Name;
    var totalItems = Model.ContainedContentTypeDefinitions.Count();
}

<script asp-name="jQuery-ui" at="Foot"></script>

<div class="form-group">
    <div id="@widgetTemplatePlaceholderId" class="widget-template-placeholder widget-template-placeholder-bagpart bagpart-@partName.HtmlClassify() row mx-0" data-buildeditorurl="@Url.Action("BuildEditor", "Admin", new { area = "OrchardCore.Flows" })">
        @{
            var index = 0;
            foreach (var widget in Model.BagPart.ContentItems)
            {
                //Create a Card Shape
                dynamic contentCard = await New.ContentCard(
                    // FOR Updater
                    CollectionShape: Model,
                    //Shape Specific
                    CollectionShapeType: nameof(BagPart),
                    ContentItem: widget,
                    BuildEditor: true,
                    ParentContentType: parentContentType,
                    CollectionPartName: partName,
                    //Card Specific Properties
                    TargetId: widgetTemplatePlaceholderId,
                    Inline: false,
                    CanMove: true,
                    CanDelete: true,
                    //Input hidden
                    //Prefixes
                    HtmlFieldPrefix: htmlFieldPrefix,
                    PrefixValue: htmlFieldPrefix + '-' + (index++).ToString(),
                    PrefixesId: Html.IdFor(x => x.Prefixes),
                    PrefixesName: Html.NameFor(x => x.Prefixes),
                    //Content Types
                    ContentTypesId: Html.IdFor(x => x.ContentTypes),
                    ContentTypesName: Html.NameFor(x => x.ContentTypes),
                    //Content Items
                    ContentItemsId: Html.IdFor(x => x.ContentItems),
                    ContentItemsName: Html.NameFor(x => x.ContentItems)
                );
                @await DisplayAsync(contentCard)
            }
        }
    </div>

    @if (totalItems == 1)
    {
        var type = Model.ContainedContentTypeDefinitions.First();

        <a class="btn btn-primary add-widget btn-sm"
           data-target-id="@widgetTemplatePlaceholderId"
           data-html-field-prefix="@htmlFieldPrefix"
           data-prefixes-name="@Html.NameFor(x => x.Prefixes)"
           data-contenttypes-name="@Html.NameFor(x => x.ContentTypes)"
           data-contentitems-name="@Html.NameFor(x => x.ContentItems)"
           data-widget-type="@type.Name"
           data-flowmetadata="false"
           data-parent-content-type="@parentContentType"
           data-part-name="@partName"
           href="javascript:;">@type.DisplayName</a>
    }
    else if (totalItems > 1)
    {
        <div class="btn-group btn-bag-add">
            <button type="button" class="btn btn-primary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                @T["Add Item"]
            </button>
            <div class="dropdown-menu">
                @foreach (var type in Model.ContainedContentTypeDefinitions)
                {
                    <a class="dropdown-item add-widget  btn-sm"
                       data-target-id="@widgetTemplatePlaceholderId"
                       data-html-field-prefix="@htmlFieldPrefix"
                       data-prefixes-name="@Html.NameFor(x => x.Prefixes)"
                       data-contenttypes-name="@Html.NameFor(x => x.ContentTypes)"
                       data-contentitems-name="@Html.NameFor(x => x.ContentItems)"
                       data-widget-type="@type.Name"
                       data-flowmetadata="false"
                       data-parent-content-type="@parentContentType"
                       data-part-name="@partName"
                       href="javascript:;">@type.DisplayName</a>
                }
            </div>
        </div>
    }

    @* Rendered only once in case multiple BagPart (or FlowPart) are on the page *@
    @if (Context.Items["BagPart.Edit"] == null)
    {
        Context.Items["BagPart.Edit"] = new object();
        <script asp-src="~/OrchardCore.Flows/Scripts/flows.edit.min.js"
                debug-src="~/OrchardCore.Flows/Scripts/flows.edit.js" at="Foot" depends-on="admin"></script>
        <style asp-src="~/OrchardCore.Widgets/Styles/widgetslist.edit.min.css"
               debug-src="~/OrchardCore.Widgets/Styles/widgetslist.edit.css"></style>
        <style asp-src="~/OrchardCore.Flows/Styles/flows.edit.min.css"
               debug-src="~/OrchardCore.Flows/Styles/flows.edit.css"></style>
    }

    @* Rendered only once per type to initialize the scripts its editor will need *@
    @foreach (var type in Model.ContainedContentTypeDefinitions)
    {
        @if (Context.Items["BagPart.Edit:" + type.Name] == null)
        {
            Context.Items["BagPart.Edit:" + type.Name] = new object();

            // Render a mock widget so that its resources are included in the page
            var contentItem = await ContentManager.NewAsync(type.Name);
            await DisplayAsync(await ContentItemDisplayManager.BuildEditorAsync(contentItem, Model.Updater, true, "", Guid.NewGuid().ToString("n")));
        }
    }

    <script at="Foot">
    @*Only apply sortable if element is present in dom, do not call sortable on scripts for mock widgets *@
    if ($('#@widgetTemplatePlaceholderId').length == 1) {
        $("#@widgetTemplatePlaceholderId").sortable({
            handle: ".widget-editor-handle",
            cursor: "move",
            opacity: 0.7,
            tolerance: "pointer",
            placeholder: {
                element: function(currentItem) {
                    return $('<div class="widget-dragdrop-placeholder"><div></div></div>')[0];
                },
                update: function (container, p) {
                    return;
                }
            },
            toleranceElement: '> div',
            start: function (event, ui) {
                widgetDragItem = ui.item;
                lastContainer = ui.item.parent();
                lastContainer.toggleClass('widget-dropzone-hint');
                widgetDragItem.find('.card').first().toggleClass('border-primary');
                var classList = widgetDragItem.attr('class').split(' ');
                $.each(classList, function (id, item) {
                    if (item.indexOf('col-md-') === 0) ui.placeholder.addClass(item);
                });
            },
            stop: function (event, ui) {
                lastContainer.toggleClass('widget-dropzone-hint');
                widgetDragItem.find('.card').first().toggleClass('border-primary');
                //clear the global variables
                widgetDragItem = null;
                lastContainer = null;
                widgetItemSourceId = null;
                widgetItemDestId = null;
                $(document).trigger('contentpreview:render');
            },
            over: function (event, ui) {
                var $self = $(this);
                if (lastContainer[0] != this) {
                    lastContainer.toggleClass('widget-dropzone-hint');
                    lastContainer = $self;
                    $self.toggleClass('widget-dropzone-hint');
                }
            }
        });
    }
    </script>
</div>
```

### LINKED ISSUE #11254: Honor secured items inside a BagPart
### Is your feature request related to a problem? Please describe.
I want some users to be able to add content-items to a specific bag while others to only be able to view only existing items.

### Describe the solution you'd like
Currently, any content item contained in a `BagPart` is always editable regardless if the contained item is securable or not. In most cases this is fine. However, I think that this function should also be extended to honor the security rules when the contained item is securable. This means, a user with `EditContent` permission can add/edit items to the `BagPart` whereas user with `ViewContent` permission only should see the content item as read only version "like preview" but should not be able to make any change to it.

### Describe alternatives you've considered
override the existing `BagPartDisplayDriver`, `ContentCard.Frame.cshtml` and `BagPart.Edit.cshtml` to do the above.

### PR DESCRIPTION
**Pull request by CrestApps**
Enhance the `BagPart` to honor permissions for the contained content types.
 
Fix #11254
Also, this would partially fix #10986 "it does not provide a clone function"

The only outstanding issue is that ordering has a slight problem. Since we don't store the position of each content item in `BagPart.ContentItems` its hard to control the order of a readonly content type. One way to fix this is to add hidden `Order` input to every content item added to a BagPart "Even when the content is in "Preview" mode.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Flows/Views/BagPart.Edit.cshtml [hunk 1]
```diff
@@ -1,18 +1,42 @@
...
 @inject OrchardCore.ContentManagement.Display.IContentItemDisplayManager ContentItemDisplayManager
-
+@inject OrchardCore.ContentManagement.IContentManager ContentManager
 @{
     var widgetTemplatePlaceholderId = Html.Id("widgetTemplatePlaceholder");
     var htmlFieldPrefix = ViewContext.ViewData.TemplateInfo.HtmlFieldPrefix;
     var parentContentType = Model.BagPart.ContentItem.ContentType;
     string partName = ((dynamic)Model).Metadata.Name;
+
+    var accessibleContentTypes = new List<ContentTypeDefinition>();
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Flows/Views/BagPart.Edit.cshtml:20]: This added block of code should be in the Driver.

**Comment by CrestApps** [comment]
@Skrypt , @sebastienros can we able to review this PR? It's a much needed enhancement for a project of mine.

**Review by Skrypt — COMMENTED** [review]
The only thing I don't like is all this code required in the View.
It should be eventually cleaned up by using the ViewModel if possible.

**Comment by hishamco** [comment]
Totally agree with @Skrypt, let us clean up the code and use ViewModel if it's possible

**Comment by CrestApps** [comment]
@Skrypt is there anything particular that you like to see moved out? I am just looking at that view now, I don't see much that could be moved to the ViewModel. Based on your last review, I moved some of the logic into the ViewModel.

**Comment by sebastienros** [comment]
> The only thing I don't like is all this code required in the View.

You could enhance the view model with all content types and their permissions already computed in the controller. `{ string ContentType, string Editable, ...}`. Such that there is less code in the view.

**Comment by CrestApps** [comment]
> > The only thing I don't like is all this code required in the View.
> You could enhance the view model with all content types and their permissions already computed in the controller. `{ string ContentType, string Editable, ...}`. Such that there is less code in the view.

@sebastienros, @Skrypt, @hishamco this was done. let me know if you see other ways to improve the code.

**Review by Skrypt — APPROVED** [review]

**Pull request merged**